### PR TITLE
GH-114 increase coverage in gateway package

### PR DESCRIPTION
**Pull request by makslion**
## What
Coverage change for the gateway package to 90.1%
Extended unit test coverage and created test files for booth controllers. 
No work on the integration right now. 

## Verification
Verification is straightforward - tests are passing on the PR and coe looks neat and covers all the cases it should - PR is verified.

### CODE REVIEW

**Comment by makslion** [comment]
/hold 
This PR now will wait for #162 and #175 and they will change a lot about the `Gateway` package and tests would need to be done over again

**Comment by jasonmadigan** [comment]
At a skim, this looks OK to me - but it would be good to get another once-over from someone who's familiar with our tests generally as well @mikenairn @david-martin @philbrookes

**Review by jasonmadigan — APPROVED** [review]

File: pkg/controllers/gateway/test_helper.go
Comment (mikenairn) [inline pkg/controllers/gateway/test_helper.go:212]: This file is being included as part of the pkg and not a test file, so it's skewing your coverage report. You should probably move it to a `test` package, or fgure out how you can include test helpers for unit tests.

Comment (makslion) [reply pkg/controllers/gateway/test_helper.go:212]: well, in fact, it bumps coverage up right now :laughing:  but I see the point. 
Why don't we take it even a bit further - I moved all that is not the test cases into the `test` package and once integration tests are done they'd end up in the same `test` package?

File: test/mockInterfaces.go
Comment (mikenairn) [inline test/mockInterfaces.go:155]: Why is this making calls via the client. You are using this in unit tests right? So it shouldn't be making calls to get k8s resources via that API.

There might be a need to mock this interface, but there shouldn't be this much code in it and the values that are going to be returned for each function should be explicit. Your tests could easy fail because of bugs in your test code.

File: test/mockInterfaces.go
Comment (mikenairn) [inline test/mockInterfaces.go:199]: This ones better, but i think it needs to be more explicit where it's being used what values are going to actually be returned if you are trying to make this into a reusable interface that any test could use where they want to mimic the placement functionality but not actually need to use it.

I had to create a similar one here https://github.com/Kuadrant/multicluster-gateway-controller/pull/237/files#diff-525abdb709ea5f8ab96976a27ec21c0e9545dd47dd82b0e2abff7a475f9672e9R40

File: test/helper.go
Comment (mikenairn) [inline test/helper.go:65]: Instead of Get, Build might be a better name for these since you are building a new instance of a resource. Get suggests you are getting it via the API.

I'd also suggest if you want these to be reusable outside of your tests that you add params to a function to let you change appropriate resource values. 

```
func BuildValidTestRequest() ctrl.Request {
	return BuildTestRequest(TestDummyCRName, TestNamespace)
}

func BuildTestRequest(name, ns string) ctrl.Request {
	return ctrl.Request{
		NamespacedName: types.NamespacedName{
			Name:      name,
			Namespace: ns,
		},
	}
}
```

File: test/helper.go
Comment (mikenairn) [inline test/helper.go:266]: Why is this here? Is this not code that belongs in the pkg?

File: test/helper.go
Comment (mikenairn) [inline test/helper.go:249]: Same here, what is this in the test package for?

File: test/mockInterfaces.go
Comment (mikenairn) [reply test/mockInterfaces.go:155]: It's being created with a fake client in the unit tests, so can ignore this comment

File: test/helper.go
Comment (makslion) [reply test/helper.go]: it is gone now - reusing one from the gateway package

File: test/helper.go
Comment (makslion) [reply test/helper.go]: reusing one from the gateway package

File: test/helper.go
Comment (makslion) [reply test/helper.go:45]: renamed to `build...`

File: test/mockInterfaces.go
Comment (makslion) [reply test/mockInterfaces.go]: resides now in the `eventhandler_test.go`

**Comment by makslion** [comment]
resolved the merge conflict and moved things around a bit. The goal is to have the `test` package hold interface implementations and functions that could be reused in other packages. The package-specific helpers now reside at the bottom of the appropriate unit test. Added build flags. 
The conversation remains about what exactly should be covered by unit and integration but this should be as close as we could get right now

File: pkg/controllers/gateway/cluster_eventhandler_test.go
Comment (mikenairn) [inline pkg/controllers/gateway/cluster_eventhandler_test.go:16]: Probably better to leave this to avoid potential confusion with our own local apis

File: pkg/controllers/gateway/gateway_controller_test.go
Comment (mikenairn) [inline pkg/controllers/gateway/gateway_controller_test.go:666]: You need to rebase main, this no longer exists in the gateway controller.

File: test/helper.go
Comment (mikenairn) [inline test/helper.go:1]: You'll probably want to rebase and make sure this still makes sense here after the e2e tests PR got merged. Also, i think you will need build tags on this

File: pkg/controllers/gateway/gateway_controller_test.go
Comment (makslion) [reply pkg/controllers/gateway/gateway_controller_test.go:666]: And a few more tests cases not don't make much sense. Removed

File: pkg/controllers/gateway/cluster_eventhandler_test.go
Comment (makslion) [reply pkg/controllers/gateway/cluster_eventhandler_test.go:16]: renamed

File: test/helper.go
Comment (makslion) [reply test/helper.go:1]: moved files around. Removed duplicate helper function. Added build tags.

File: pkg/controllers/gateway/params_test.go
Comment (mikenairn) [inline pkg/controllers/gateway/params_test.go:89]: I think you can get rid of the addr function and use testutil.Pointer everywhere?

File: test/util/helper.go
Comment (mikenairn) [inline test/util/helper.go:24]: FailFetchDNSSubdomain

File: pkg/controllers/gateway/params_test.go
Comment (makslion) [reply pkg/controllers/gateway/params_test.go:89]: missed one! I knew those sneaky calls would get me. Done

File: pkg/controllers/gateway/gateway_controller_test.go
Comment (mikenairn) [inline pkg/controllers/gateway/gateway_controller_test.go:162]: gatewayckasses?

File: pkg/controllers/gateway/gateway_controller_test.go
Comment (mikenairn) [inline pkg/controllers/gateway/gateway_controller_test.go:164]: classm?

File: pkg/controllers/gateway/gateway_controller_test.go
Comment (makslion) [reply pkg/controllers/gateway/gateway_controller_test.go:162]: the funny part as it was passing with the typo. Fixed

File: pkg/controllers/gateway/gateway_controller_test.go
Comment (makslion) [reply pkg/controllers/gateway/gateway_controller_test.go:164]: fixed

**Pull request merged**
